### PR TITLE
meson: Re-enable test binaries

### PR DESCRIPTION
**Pull request by chouquette**

### CODE REVIEW

File: config.h.meson [hunk 1]
```diff
@@ -136,7 +136,7 @@
 #mesondefine URANDOM
 
 /* Version number of package */
-#mesondefine VERSION
+#define VERSION "@VERSION@"
```
Comment (eli-schwartz) [inline config.h.meson:139]: I think you just want to use `conf_data.set_quoted` and leave this a `#mesondefine`.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: meson.build [hunk 2]
```diff
@@ -280,6 +281,10 @@ else
   warning('Doxygen not found - continuing without Doxygen support')
 endif
 
+if not get_option('binaries').disabled()
```
Comment (eli-schwartz) [inline meson.build:284]: Why not just use a boolean option for binaries? Why does it need to be a tristate feature if you aren't using all 3 possible values to do 3 different things?

In other words, what's the difference between `-Dbinaries=enabled` and `-Dbinaries=auto`?

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: config.h.meson (hunk 1 above)
Comment (chouquette) [reply config.h.meson:139]: Ah indeed, I completely missed the `set_quoted` variant. Fixed locally, thanks!

File: meson.build (hunk 2 above)
Comment (eli-schwartz) [reply meson.build:284]: The original purpose for feature options is that for functions (dependency, find_library...) that have a `required: ...` kwarg, features do something a bit clever.

=enabled acts like boolean true, but *auto* behaves like `dependency('foo', required: false)` (check for it but don't require it, and =disabled behaves like an `if` statement but with better logging: it is comparable to `if opt -> dependency('foo', required: true)`.

So a feature option has 3 states and is special-cased by some functions, such that:
- enabled means required
- auto means use an [automagic dependency](https://wiki.gentoo.org/wiki/Project:Quality_Assurance/Automagic_dependencies) (autotools would call this --enable-foo=check)
- disabled means log "dependency foo skipped: feature binaries disabled"

The auto-features packaging option is an extension of this concept which allows distro packagers to automatically force automagic dependencies to resolve one way or the other, unless overridden later on in the Meson command line. The reason why it is so useful is not to "add optional features" but rather to "control automatic probing for system dependencies that may or may not be installed, which unpredictably change the results of the build".

Considering that these test binaries are:
- not installed
- free of dependencies other than libdsm

I personally think there is no harm in unconditionally building them no matter what. Why even bother with an option?

If they take a long time to build, a better option might be marking them to not `build_by_default`, i.e. they would not build by `ninja` and only be built if the user runs `ninja bin/dsm` or `ninja bin/discover`.

Comment (chouquette) [reply meson.build]: Thanks a lot for the details explanation!

Regarding the option presence, it's often not about the time it takes but rather some shortcuts taken by the test binaries, which often don't work in some constrained environment such as UWP/Android/iOS due to lack of some features.

I switched back to a boolean

Comment (eli-schwartz) [reply meson.build]: Even in that case, `build_by_default: false` would mean that a bare `ninja` doesn't error out I guess. But fair enough.

**Review by jbkempf — APPROVED** [review]

**Comment by jbkempf** [comment]
LGTM

**Pull request merged**